primitives/staking/src/lib.rs

```diff
@@ -69,10 +75,62 @@ pub struct Stake<T: StakingInterface + ?Sized> {
 	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
 	/// non-staking locks on the bonded pair's balance this amount is going to be larger in
 	/// reality.
-	pub total: T::Balance,
+	pub total: Balance,
 	/// The total amount of the stash's balance that will be at stake in any forthcoming
 	/// rounds.
-	pub active: T::Balance,
+	pub active: Balance,
+}
+
+/// A generic staking event listener.
+/// Note that the interface is designed in a way that the events are fired post-action, so any
+/// pre-action data that is needed needs to be passed to interface methods.
+/// The rest of the data can be retrieved by using `StakingInterface`.
+pub trait OnStakingUpdate<AccountId: Clone, Balance: Copy> {
+	/// Fired when the stake amount of someone updates.
+	///
+	/// Also called when someone stakes for the first time. (TODO: is it? this is why we need unit
+	/// tests for this pallet alone).
+	///
+	/// This is effectively any changes to the bond amount, such as bonding more funds, and
+	/// unbonding.
+	fn on_stake_update(who: &AccountId, prev_stake: Option<Stake<AccountId, Balance>>);
+	/// Fired when someone sets their intention to nominate, either new, or existing one.
+	fn on_nominator_update(who: &AccountId, prev_nominations: Vec<AccountId>);
+	/// Fired when someone sets their intention to validate, either new, or existing one.
+	fn on_validator_update(who: &AccountId);
+	/// Fired when someone removes their intention to validate, either due to chill or nominating.
+	fn on_validator_remove(who: &AccountId); // only fire this event when this is an actual Validator
+	/// Fired when someone removes their intention to nominate, either due to chill or validating.
+	fn on_nominator_remove(who: &AccountId, nominations: Vec<AccountId>); // only fire this if this is an actual Nominator
+	/// fired when someone is fully unstaked.
+	fn on_unstake(who: &AccountId); // -> basically `kill_stash`
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(10)]
```

Comment:
This should work fine with the default derive (the simplest version in the documentation of `impl_trait_for_tuples`)

Reply:
The documentation for impl_trait_for_tuples is mostly mentioning `impl_for_tuples(5)` and clearly states that this can be called with one or two params.

Reply:
I guess 10 is a reasonable default, but if you want to make it a 5 - not a problem  for me too.

Reply:
I am saying that if you put this on top of the trait, it would be enough.